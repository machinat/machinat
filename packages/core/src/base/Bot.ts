import type { SociablyBot, SociablyThread, SociablyNode } from '../types.js';
import type { DispatchResponse } from '../engine/types.js';
import { serviceInterface, serviceProviderClass } from '../service/index.js';

type BaseBotI = SociablyBot<SociablyThread, unknown, unknown>;

const BotPlatformMapI = serviceInterface<BaseBotI>({
  name: 'BotPlatformMap',
  polymorphic: true,
});

/**
 * @category Base
 */
export class BaseBot implements BaseBotI {
  static PlatformMap = BotPlatformMapI;

  private _platformMapping: Map<string, BaseBotI>;

  constructor(platformMapping: Map<string, BaseBotI>) {
    this._platformMapping = platformMapping;
  }

  async render(
    thread: SociablyThread,
    node: SociablyNode
  ): Promise<null | DispatchResponse<unknown, unknown>> {
    const bot = this._platformMapping.get(thread.platform);
    if (!bot) {
      throw new TypeError(
        `thread of platform '${thread.platform}' is not supported`
      );
    }

    return bot.render(thread, node);
  }

  async start(): Promise<void> {} // eslint-disable-line class-methods-use-this
  async stop(): Promise<void> {} // eslint-disable-line class-methods-use-this
}

const BotP = serviceProviderClass({
  lifetime: 'transient',
  deps: [BotPlatformMapI],
})(BaseBot);

type BotP = BaseBot;

export default BotP;
