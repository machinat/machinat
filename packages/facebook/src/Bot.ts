import { AgentSettingsAccessor } from '@sociably/core';
import Engine, { DispatchError } from '@sociably/core/engine';
import Queue from '@sociably/core/queue';
import Renderer from '@sociably/core/renderer';
import ModuleUtilitiesI from '@sociably/core/base/ModuleUtilities';
import { serviceProviderClass } from '@sociably/core/service';
import type {
  SociablyNode,
  SociablyBot,
  InitScopeFn,
  DispatchWrapper,
} from '@sociably/core';
import {
  MetaApiWorker,
  MetaApiJob,
  MetaApiResult,
  MetaApiDispatchResponse,
  MetaApiResponseBody,
} from '@sociably/meta-api';
import generalComponentDelegator from './components/general';
import { FACEBOOK, PATH_FEED, PATH_PHOTOS } from './constant';
import {
  ConfigsI,
  PlatformUtilitiesI,
  PageSettingsAccessorI,
} from './interface';
import FacebookPage from './Page';
import FacebookChat from './Chat';
import InteractTarget from './InteractTarget';
import {
  createChatJobs,
  createChatAttachmentJobs,
  createPostJobs,
  createInteractJobs,
} from './job';
import type {
  FacebookThread,
  FacebookComponent,
  FacebookSegmentValue,
  FacebookDispatchFrame,
  MessagingOptions,
  FacebookPageSettings,
} from './types';

type FacebookDispatchTarget = FacebookThread | FacebookPage;

type FacebookBotOptions = {
  appSecret?: string;
  graphApiVersion?: string;
  apiBatchRequestInterval?: number;
  pageSettingsAccessor: AgentSettingsAccessor<
    FacebookPage,
    FacebookPageSettings
  >;
  initScope?: InitScopeFn;
  dispatchWrapper?: DispatchWrapper<
    MetaApiJob,
    FacebookDispatchFrame,
    MetaApiResult
  >;
};

type UploadAttachmentResult = {
  attachmentId: string;
};

type PagePhotoResult = {
  photoId: string;
};

type PagePostResult = {
  postId: string;
  photos: PagePhotoResult[];
};

type CommentResult = {
  commentId: string;
  photo: null | PagePhotoResult;
};

type ApiCallOptions = {
  /** The page to make the API call */
  page: string | FacebookPage;
  /** HTTP method */
  method?: string;
  /** API request URL relative to https://graph.facebook.com/{version}/ */
  url: string;
  /** API request parameters */
  params?: Record<string, unknown>;
};

/**
 * FacebookBot render messages and make API call to Facebook platform.
 * @category Provider
 */
export class FacebookBot
  implements SociablyBot<FacebookThread, MetaApiJob, MetaApiResult>
{
  worker: MetaApiWorker;
  engine: Engine<
    FacebookDispatchTarget,
    FacebookSegmentValue,
    FacebookComponent<unknown>,
    MetaApiJob,
    MetaApiResult
  >;

  platform = FACEBOOK;

  constructor({
    appSecret,
    graphApiVersion = 'v11.0',
    apiBatchRequestInterval = 500,
    pageSettingsAccessor,
    initScope,
    dispatchWrapper,
  }: FacebookBotOptions) {
    const renderer = new Renderer<
      FacebookSegmentValue,
      FacebookComponent<unknown>
    >(FACEBOOK, generalComponentDelegator);

    const queue = new Queue<MetaApiJob, MetaApiResult>();
    const worker = new MetaApiWorker(
      pageSettingsAccessor,
      appSecret,
      graphApiVersion,
      apiBatchRequestInterval
    );

    this.engine = new Engine(
      FACEBOOK,
      renderer,
      queue,
      worker,
      initScope,
      dispatchWrapper
    );
  }

  async start(): Promise<void> {
    this.engine.start();
  }

  async stop(): Promise<void> {
    this.engine.stop();
  }

  async render(
    target: FacebookDispatchTarget,
    node: SociablyNode
  ): Promise<null | MetaApiDispatchResponse> {
    if (target instanceof FacebookChat) {
      return this.engine.render(target, node, createChatJobs());
    }
    if (target instanceof FacebookPage) {
      return this.engine.render(target, node, createPostJobs);
    }
    if (target instanceof InteractTarget) {
      return this.engine.render(target, node, createInteractJobs);
    }
    throw new TypeError('invalid rendering target');
  }

  /** Send messages or actions on a chat */
  async message(
    chat: FacebookChat,
    messages: SociablyNode,
    options?: MessagingOptions
  ): Promise<null | MetaApiDispatchResponse> {
    return this.engine.render(chat, messages, createChatJobs(options));
  }

  /** Upload a media chat attachment for later use */
  async uploadChatAttachment(
    /** The {@link FacebookPage} that owns the attachment */
    pageInput: string | FacebookPage,
    /** An {@link Image}, {@link Audio}, {@link Video} or {@link File} element to be uploaded */ node: SociablyNode
  ): Promise<null | UploadAttachmentResult> {
    const page =
      typeof pageInput === 'string' ? new FacebookPage(pageInput) : pageInput;
    const response = await this.engine.render(
      page,
      node,
      createChatAttachmentJobs
    );
    const result = response?.results[0].body;
    return result ? { attachmentId: result.attachment_id } : null;
  }

  /** Create a post or a photo on the page feed */
  async post(
    /** The {@link FacebookPage} to post */
    pageInput: string | FacebookPage,
    /** Text, a {@link PagePost} or a {@link PagePhoto} to post */
    node: SociablyNode
  ): Promise<null | PagePostResult> {
    const page =
      typeof pageInput === 'string' ? new FacebookPage(pageInput) : pageInput;
    const response = await this.engine.render(page, node, createPostJobs);
    if (!response) {
      return null;
    }

    const photos: PagePhotoResult[] = [];
    let postId: undefined | string;

    const { jobs, results } = response;
    for (const [i, { request }] of jobs.entries()) {
      const resultBody = results[i].body;

      if (request.url === PATH_FEED) {
        postId = resultBody.id;
      } else if (request.url === PATH_PHOTOS) {
        if (resultBody.post_id) {
          postId = resultBody.post_id;
        }
        photos.push({ photoId: resultBody.id });
      }
    }
    if (!postId) {
      return null;
    }
    return { postId, photos };
  }

  /** Comment or make reactions to a Facebook post or comment */
  async interact(
    /** The target thread to interact with */
    target: InteractTarget,
    /** {@link Comment} or {@link Reaction} */
    node: SociablyNode
  ): Promise<null | CommentResult[]> {
    const response = await this.engine.render(target, node, createInteractJobs);
    if (!response) {
      return null;
    }

    const results: CommentResult[] = [];
    let photo: null | PagePhotoResult = null;

    for (const [i, { request }] of response.jobs.entries()) {
      if (request.url === PATH_PHOTOS) {
        photo = { photoId: response.results[i].body.id };
      } else if (request.url === PATH_PHOTOS) {
        results.push({
          commentId: response.results[i].body.id,
          photo,
        });
        photo = null;
      }
    }
    return results;
  }

  async requestApi<ResBody extends MetaApiResponseBody>({
    page: pageInput,
    method = 'GET',
    url,
    params,
  }: ApiCallOptions): Promise<ResBody> {
    const page =
      typeof pageInput === 'string' ? new FacebookPage(pageInput) : pageInput;
    try {
      const { results } = await this.engine.dispatchJobs(page, [
        {
          channel: page,
          request: { method, url, params },
        },
      ]);

      return results[0].body as ResBody;
    } catch (err) {
      if (err instanceof DispatchError) {
        throw err.errors[0];
      }
      throw err;
    }
  }
}

const BotP = serviceProviderClass({
  lifetime: 'singleton',
  deps: [
    ConfigsI,
    PageSettingsAccessorI,
    { require: ModuleUtilitiesI, optional: true },
    { require: PlatformUtilitiesI, optional: true },
  ],
  factory: (
    { appSecret, apiBatchRequestInterval },
    pageSettingsAccessor,
    moduleUitils,
    platformUtils
  ) =>
    new FacebookBot({
      pageSettingsAccessor,
      appSecret,
      apiBatchRequestInterval,
      initScope: moduleUitils?.initScope,
      dispatchWrapper: platformUtils?.dispatchWrapper,
    }),
})(FacebookBot);

type BotP = FacebookBot;
export default BotP;
