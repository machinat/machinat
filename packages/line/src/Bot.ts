import invariant from 'invariant';
import type {
  SociablyNode,
  SociablyBot,
  InitScopeFn,
  DispatchWrapper,
} from '@sociably/core';
import Renderer from '@sociably/core/renderer';
import Queue from '@sociably/core/queue';
import Engine, { DispatchError } from '@sociably/core/engine';
import ModuleUtilitiesI from '@sociably/core/base/ModuleUtilities';
import { makeClassProvider } from '@sociably/core/service';

import { createChatJobs, createMulticastJobs } from './job';
import generalElementDelegate from './components/general';
import LineWorker from './Worker';
import LineChat from './Chat';
import { ConfigsI, PlatformUtilitiesI } from './interface';
import { LINE } from './constant';
import type {
  LineSource,
  LineSegmentValue,
  LineComponent,
  LineJob,
  LineResult,
  LineDispatchFrame,
  LineDispatchResponse,
  MessagingApiResult,
} from './types';

type LineBotOptions = {
  providerId: string;
  channelId: string;
  accessToken: string;
  maxRequestConnections?: number;
  initScope?: InitScopeFn;
  dispatchWrapper?: DispatchWrapper<LineJob, LineDispatchFrame, LineResult>;
};

/**
 * @category Provider
 */
export class LineBot implements SociablyBot<LineChat, LineJob, LineResult> {
  providerId: string;
  channelId: string;
  maxRequestConnections: number;
  engine: Engine<
    LineChat,
    LineSegmentValue,
    LineComponent<unknown>,
    LineJob,
    LineResult
  >;

  platform = LINE;

  constructor({
    providerId,
    channelId,
    accessToken,
    maxRequestConnections = 100,
    initScope,
    dispatchWrapper,
  }: LineBotOptions) {
    invariant(accessToken, 'configs.accessToken should not be empty');
    invariant(providerId, 'configs.providerId should not be empty');
    invariant(channelId, 'configs.channelId should not be empty');

    this.providerId = providerId;
    this.channelId = channelId;
    this.maxRequestConnections = maxRequestConnections;

    const queue = new Queue<LineJob, LineResult>();
    const worker = new LineWorker(accessToken, maxRequestConnections);
    const renderer = new Renderer<LineSegmentValue, LineComponent<unknown>>(
      LINE,
      generalElementDelegate
    );

    this.engine = new Engine(
      LINE,
      renderer,
      queue,
      worker,
      initScope,
      dispatchWrapper
    );
  }

  async start(): Promise<void> {
    this.engine.start();
  }

  async stop(): Promise<void> {
    this.engine.stop();
  }

  render(
    source: string | LineSource | LineChat,
    message: SociablyNode,
    options?: { replyToken?: string }
  ): Promise<null | LineDispatchResponse> {
    const thread =
      source instanceof LineChat
        ? source
        : typeof source === 'string'
        ? new LineChat(this.channelId, 'user', source)
        : LineChat.fromMessagingSource(this.channelId, source);

    return this.engine.render(
      thread,
      message,
      createChatJobs(options && options.replyToken)
    );
  }

  renderMulticast(
    targets: string[],
    message: SociablyNode
  ): Promise<null | LineDispatchResponse> {
    return this.engine.render(null, message, createMulticastJobs(targets));
  }

  async makeApiCall<ResBody extends MessagingApiResult>(
    method: 'GET' | 'POST' | 'PUT' | 'DELETE',
    path: string,
    body?: unknown
  ): Promise<ResBody> {
    try {
      const response = await this.engine.dispatchJobs(null, [
        { method, path, body, executionKey: undefined },
      ]);

      return response.results[0].body as ResBody;
    } catch (err) {
      if (err instanceof DispatchError) {
        throw err.errors[0];
      }
      throw err;
    }
  }
}

const BotP = makeClassProvider({
  lifetime: 'singleton',
  deps: [
    ConfigsI,
    { require: ModuleUtilitiesI, optional: true },
    { require: PlatformUtilitiesI, optional: true },
  ],
  factory: (
    { providerId, channelId, accessToken, maxRequestConnections },
    moduleUtils,
    platformUtils
  ) =>
    new LineBot({
      providerId,
      channelId,
      accessToken,
      maxRequestConnections,
      initScope: moduleUtils?.initScope,
      dispatchWrapper: platformUtils?.dispatchWrapper,
    }),
})(LineBot);

type BotP = LineBot;

export default BotP;
