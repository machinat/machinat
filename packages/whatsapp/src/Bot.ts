import invariant from 'invariant';
import type {
  SociablyNode,
  SociablyBot,
  SociablyChannel,
  InitScopeFn,
  DispatchWrapper,
  ChannelSettingsAccessor,
} from '@sociably/core';
import Engine, { DispatchError } from '@sociably/core/engine';
import Queue from '@sociably/core/queue';
import Renderer from '@sociably/core/renderer';
import ModuleUtilitiesI from '@sociably/core/base/ModuleUtilities';
import { makeClassProvider } from '@sociably/core/service';
import {
  MetaApiWorker,
  MetaApiJob,
  MetaApiResult,
  MetaApiDispatchResponse,
  MetaApiResponseBody,
} from '@sociably/meta-api';
import generalComponentDelegator from './components/general';
import { WHATSAPP } from './constant';
import { ConfigsI, PlatformUtilitiesI } from './interface';
import WhatsAppChat from './Chat';
import WhatsAppAgent from './Agent';
import { createChatJobs, createUploadingMediaJobs } from './job';
import type {
  WhatsAppComponent,
  WhatsAppSegmentValue,
  WhatsAppDispatchFrame,
} from './types';

type WhatsAppBotOptions = {
  accessToken: string;
  appSecret?: string;
  graphApiVersion?: string;
  apiBatchRequestInterval?: number;
  initScope?: InitScopeFn;
  dispatchWrapper?: DispatchWrapper<
    MetaApiJob,
    WhatsAppDispatchFrame,
    MetaApiResult
  >;
};

const DUMMY_API_CALL_CHANNEL: SociablyChannel = {
  platform: WHATSAPP,
  uid: 'whatsapp:dummy_api_call_channel',
  uniqueIdentifier: {
    platform: WHATSAPP,
    id: 'dummy_api_call_channel',
  },
};

const createStaticSettingsAccessor = (
  accessToken: string
): ChannelSettingsAccessor<WhatsAppAgent, { accessToken: string }> => ({
  getChannelSettings: async () => ({ accessToken }),
  getChannelSettingsBatch: async (channels) =>
    channels.map(() => ({ accessToken })),
  listAllChannelSettings: async () => [{ accessToken }],
});

type ApiCallOptions = {
  method: 'GET' | 'POST' | 'PUT' | 'DELETE';
  path: string;
  params?: Record<string, unknown>;
};

/**
 * WhatsAppBot render messages and make API call to WhatsApp platform.
 * @category Provider
 */
export class WhatsAppBot
  implements SociablyBot<WhatsAppChat, MetaApiJob, MetaApiResult>
{
  worker: MetaApiWorker;
  engine: Engine<
    null | WhatsAppChat | WhatsAppAgent,
    WhatsAppSegmentValue,
    WhatsAppComponent<unknown>,
    MetaApiJob,
    MetaApiResult
  >;

  platform = WHATSAPP;

  constructor({
    accessToken,
    appSecret,
    graphApiVersion = 'v11.0',
    apiBatchRequestInterval = 500,
    initScope,
    dispatchWrapper,
  }: WhatsAppBotOptions) {
    invariant(accessToken, 'options.accessToken should not be empty');

    const renderer = new Renderer<
      WhatsAppSegmentValue,
      WhatsAppComponent<unknown>
    >(WHATSAPP, generalComponentDelegator);

    const queue = new Queue<MetaApiJob, MetaApiResult>();
    const worker = new MetaApiWorker(
      createStaticSettingsAccessor(accessToken),
      appSecret,
      graphApiVersion,
      apiBatchRequestInterval
    );

    this.engine = new Engine(
      WHATSAPP,
      renderer,
      queue,
      worker,
      initScope,
      dispatchWrapper
    );
  }

  async start(): Promise<void> {
    this.engine.start();
  }

  async stop(): Promise<void> {
    this.engine.stop();
  }

  async render(
    thread: WhatsAppChat,
    messages: SociablyNode
  ): Promise<null | MetaApiDispatchResponse> {
    return this.engine.render(thread, messages, createChatJobs);
  }

  async uploadMedia(
    agent: WhatsAppAgent,
    node: SociablyNode
  ): Promise<null | { id: string }> {
    const response = await this.engine.render(
      agent,
      node,
      createUploadingMediaJobs
    );

    if (!response) {
      return null;
    }
    return response.results[0].body as { id: string };
  }

  async makeApiCall<ResBody extends MetaApiResponseBody>({
    method,
    path,
    params,
  }: ApiCallOptions): Promise<ResBody> {
    try {
      const { results } = await this.engine.dispatchJobs(null, [
        {
          channel: DUMMY_API_CALL_CHANNEL,
          request: {
            method,
            relativeUrl: path,
            params,
          },
        },
      ]);

      return results[0].body as ResBody;
    } catch (err) {
      if (err instanceof DispatchError) {
        throw err.errors[0];
      }
      throw err;
    }
  }
}

const BotP = makeClassProvider({
  lifetime: 'singleton',
  deps: [
    ConfigsI,
    { require: ModuleUtilitiesI, optional: true },
    { require: PlatformUtilitiesI, optional: true },
  ],
  factory: (
    { accessToken, appSecret, apiBatchRequestInterval },
    moduleUitils,
    platformUtils
  ) =>
    new WhatsAppBot({
      accessToken,
      appSecret,
      apiBatchRequestInterval,
      initScope: moduleUitils?.initScope,
      dispatchWrapper: platformUtils?.dispatchWrapper,
    }),
})(WhatsAppBot);

type BotP = WhatsAppBot;
export default BotP;
