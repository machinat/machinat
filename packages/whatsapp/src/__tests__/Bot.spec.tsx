import querystring from 'querystring';
import nock from 'nock';
import moxy, { Moxy } from '@moxyjs/moxy';
import Sociably from '@sociably/core';
import Queue from '@sociably/core/queue';
import _Renderer from '@sociably/core/renderer';
import _Engine from '@sociably/core/engine';
import { MetaApiWorker as _Worker, MetaApiError } from '@sociably/meta-api';
import WhatsAppAgent from '../Agent.js';
import WhatsAppChat from '../Chat.js';
import { WhatsAppBot } from '../Bot.js';
import { Image, ButtonsTemplate, ReplyButton } from '../components/index.js';

const Renderer = _Renderer as Moxy<typeof _Renderer>;
const Engine = _Engine as Moxy<typeof _Engine>;
const Worker = _Worker as Moxy<typeof _Worker>;

nock.disableNetConnect();

jest.mock('@sociably/core/engine', () =>
  moxy(jest.requireActual('@sociably/core/engine')),
);
jest.mock('@sociably/core/renderer', () =>
  moxy(jest.requireActual('@sociably/core/renderer')),
);
jest.mock('@sociably/meta-api', () =>
  moxy(jest.requireActual('@sociably/meta-api'), {
    includeProperties: ['MetaApiWorker'],
  }),
);

const initScope = moxy(() => moxy());
const dispatchWrapper = moxy((x) => x);

const businessNumber = '1234567890';
const accessToken = '_ACCESS_TOKEN_';
const appId = '_APP_ID_';
const appSecret = '_APP_SECRET_';

const message = (
  <ButtonsTemplate
    header={<Image url="http://sociably.com/foo.jpg" />}
    buttons={<ReplyButton title="BAR" data="BAZ" />}
  >
    Hello <b>World!</b>
  </ButtonsTemplate>
);

let graphApi;
const bodySpy = moxy(() => true);

beforeEach(() => {
  graphApi = nock('https://graph.facebook.com').post('/v17.0/', bodySpy);
  bodySpy.mock.clear();
  Engine.mock.clear();
  Renderer.mock.clear();
  Worker.mock.clear();
});

afterEach(() => {
  nock.cleanAll();
});

describe('#constructor(options)', () => {
  it('throw if accessToken not given', () => {
    expect(
      () => new WhatsAppBot({ businessNumber, appSecret } as never),
    ).toThrowErrorMatchingInlineSnapshot(
      `"options.accessToken should not be empty"`,
    );
  });

  it('assemble core modules', () => {
    const bot = new WhatsAppBot({
      appId,
      appSecret,
      initScope,
      dispatchWrapper,
      accessToken,
    });

    expect(bot.engine).toBeInstanceOf(Engine);

    expect(Renderer).toHaveBeenCalledTimes(1);
    expect(Renderer).toHaveBeenCalledWith('whatsapp', expect.any(Function));

    expect(Engine).toHaveBeenCalledTimes(1);
    expect(Engine).toHaveBeenCalledWith(
      'whatsapp',
      expect.any(Renderer),
      expect.any(Queue),
      expect.any(Worker),
      initScope,
      dispatchWrapper,
    );

    expect(Worker).toHaveBeenCalledTimes(1);
    expect(Worker).toHaveBeenCalledWith({
      appId,
      appSecret,
      defaultAccessToken: accessToken,
      graphApiVersion: 'v17.0',
      consumeInterval: 500,
    });
  });

  it('pass options to worker', async () => {
    expect(
      new WhatsAppBot({
        initScope,
        dispatchWrapper,
        accessToken,
        appId,
        appSecret,
        graphApiVersion: 'v8.0',
        apiBatchRequestInterval: 0,
      }),
    );

    expect(Worker).toHaveBeenCalledTimes(1);
    expect(Worker).toHaveBeenCalledWith({
      appId,
      appSecret,
      defaultAccessToken: accessToken,
      graphApiVersion: 'v8.0',
      consumeInterval: 0,
    });
  });
});

test('#start() and #stop() start/stop engine', () => {
  const bot = new WhatsAppBot({
    initScope,
    dispatchWrapper,
    accessToken,
    appId,
    appSecret,
  });

  bot.start();
  expect(bot.engine.start).toHaveBeenCalledTimes(1);

  bot.stop();
  expect(bot.engine.stop).toHaveBeenCalledTimes(1);
});

describe('#render(thread, message, options)', () => {
  const bot = new WhatsAppBot({ accessToken, appId, appSecret });
  const sucessfulResult = {
    messaging_product: 'whatsapp',
    contacts: [{ input: '9876543210', wa_id: '9876543210' }],
    messages: [{ id: 'wamid....' }],
  };

  let apiStatus;
  beforeEach(() => {
    const messageResult = {
      code: 200,
      body: JSON.stringify(sucessfulResult),
    };
    apiStatus = graphApi.reply(200, [messageResult, messageResult]);
    bot.start();
  });

  afterEach(() => {
    bot.stop();
  });

  it('resolves null if message is empty', async () => {
    const empties = [undefined, null, [], <></>];
    for (const empty of empties) {
      // eslint-disable-next-line no-await-in-loop
      await expect(
        bot.render(new WhatsAppChat('1234567890', '9876543210'), empty),
      ).resolves.toBe(null);
      expect(apiStatus.isDone()).toBe(false);
    }
  });

  it('post to NUMBER/messages api', async () => {
    const response = await bot.render(
      new WhatsAppChat('1234567890', '9876543210'),
      message,
    );
    expect(response).toMatchSnapshot();

    for (const result of response!.results) {
      expect(result).toEqual({
        code: 200,
        body: sucessfulResult,
      });
    }

    expect(bodySpy).toHaveBeenCalledTimes(1);
    const body = bodySpy.mock.calls[0].args[0];

    for (const request of JSON.parse(body.batch)) {
      expect(request.method).toBe('POST');
      expect(request.relative_url).toMatchInlineSnapshot(
        `"1234567890/messages?access_token=_ACCESS_TOKEN_&appsecret_proof=932e1d758c8379099e1b7f9e75e1abf41ab496760d64ddb05e3d21979d13c31f"`,
      );
    }

    expect(body).toMatchSnapshot();
    expect(querystring.decode(JSON.parse(body.batch)[0].body))
      .toMatchInlineSnapshot(`
      {
        "interactive": "{"type":"button","body":{"text":"Hello *World!*"},"header":{"type":"image","image":{"link":"http://sociably.com/foo.jpg"}},"action":{"buttons":[{"type":"reply","reply":{"title":"BAR","id":"BAZ"}}]}}",
        "messaging_product": "whatsapp",
        "to": "9876543210",
        "type": "interactive",
      }
    `);

    expect(apiStatus.isDone()).toBe(true);
  });
});

describe('#uploadMedia(message)', () => {
  const bot = new WhatsAppBot({ accessToken, appId, appSecret });

  beforeEach(() => {
    bot.start();
  });

  afterEach(() => {
    bot.stop();
  });

  it('resolves null if message is empty', async () => {
    const empties = [undefined, null, [], <></>];
    for (const empty of empties) {
      // eslint-disable-next-line no-await-in-loop
      await expect(
        bot.uploadMedia(new WhatsAppAgent('1234567890'), empty),
      ).resolves.toBe(null);
    }
  });

  it('post /media api', async () => {
    const apiStatus = graphApi.reply(200, [
      { code: 200, body: JSON.stringify({ id: 401759795 }) },
    ]);

    const result = await bot.uploadMedia(
      new WhatsAppAgent('1234567890'),
      <Image file={{ data: Buffer.from('foo'), contentType: 'image/png' }} />,
    );
    expect(result).toEqual({ id: 401759795 });

    expect(bodySpy).toHaveBeenCalledTimes(1);
    const body = bodySpy.mock.calls[0].args[0];

    expect(body.replace(/-----\d+/g, '-----SEPERATOR')).toMatchSnapshot();

    expect(apiStatus.isDone()).toBe(true);
  });
});

describe('#requestApi()', () => {
  it('call facebook graph api', async () => {
    const bot = new WhatsAppBot({ accessToken, appId, appSecret });
    bot.start();

    const apiCall = graphApi.reply(200, [{ code: 200, body: '{"foo":"bar"}' }]);

    await expect(
      bot.requestApi({ method: 'POST', url: 'foo', params: { bar: 'baz' } }),
    ).resolves.toEqual({
      foo: 'bar',
    });

    expect(apiCall.isDone()).toBe(true);
  });

  it('throw MetaApiError if api call fail', async () => {
    const bot = new WhatsAppBot({ accessToken, appId, appSecret });
    bot.start();

    const apiCall = graphApi.reply(200, [
      {
        code: 400,
        body: JSON.stringify({
          error: {
            message: 'bad',
            type: 'devil',
            code: 444,
            error_subcode: 666,
            fbtrace_id: 'xxxx',
          },
        }),
      },
    ]);

    try {
      await bot.requestApi({
        method: 'POST',
        url: 'foo',
        params: { bar: 'baz' },
      });
      expect('should not be here').toBeFalsy();
    } catch (err) {
      expect(err).toBeInstanceOf(MetaApiError);
      expect(err.code).toBe(444);
      expect(err.message).toBe('bad');
    }

    expect(apiCall.isDone()).toBe(true);
  });
});
