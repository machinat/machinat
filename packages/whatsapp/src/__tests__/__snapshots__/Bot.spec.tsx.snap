// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render(thread, message, options) post to NUMBER/messages api 1`] = `
Object {
  "jobs": Array [
    Object {
      "channel": WhatsAppAgent {
        "$$typeofChannel": true,
        "numberId": "1234567890",
        "platform": "whatsapp",
      },
      "consumeResult": undefined,
      "key": "wa.1234567890.9876543210",
      "request": Object {
        "method": "POST",
        "params": Object {
          "context": undefined,
          "interactive": Object {
            "actions": Object {
              "buttons": Array [
                Object {
                  "id": "BAZ",
                  "title": "BAR",
                  "type": "reply",
                },
              ],
            },
            "body": Object {
              "text": "Hello *World!*",
            },
            "footer": undefined,
            "header": Object {
              "image": Object {
                "caption": undefined,
                "filename": undefined,
                "id": undefined,
                "link": "http://sociably.com/foo.jpg",
              },
              "type": "image",
            },
            "type": "buttons",
          },
          "messaging_product": "whatsapp",
          "to": "9876543210",
          "type": "interactive",
        },
        "url": "1234567890/messages",
      },
    },
  ],
  "results": Array [
    Object {
      "body": Object {
        "contacts": Array [
          Object {
            "input": "9876543210",
            "wa_id": "9876543210",
          },
        ],
        "messages": Array [
          Object {
            "id": "wamid....",
          },
        ],
        "messaging_product": "whatsapp",
      },
      "code": 200,
    },
  ],
  "tasks": Array [
    Object {
      "payload": Array [
        Object {
          "channel": WhatsAppAgent {
            "$$typeofChannel": true,
            "numberId": "1234567890",
            "platform": "whatsapp",
          },
          "consumeResult": undefined,
          "key": "wa.1234567890.9876543210",
          "request": Object {
            "method": "POST",
            "params": Object {
              "context": undefined,
              "interactive": Object {
                "actions": Object {
                  "buttons": Array [
                    Object {
                      "id": "BAZ",
                      "title": "BAR",
                      "type": "reply",
                    },
                  ],
                },
                "body": Object {
                  "text": "Hello *World!*",
                },
                "footer": undefined,
                "header": Object {
                  "image": Object {
                    "caption": undefined,
                    "filename": undefined,
                    "id": undefined,
                    "link": "http://sociably.com/foo.jpg",
                  },
                  "type": "image",
                },
                "type": "buttons",
              },
              "messaging_product": "whatsapp",
              "to": "9876543210",
              "type": "interactive",
            },
            "url": "1234567890/messages",
          },
        },
      ],
      "type": "dispatch",
    },
  ],
}
`;

exports[`#render(thread, message, options) post to NUMBER/messages api 2`] = `
Object {
  "access_token": "_ACCESS_TOKEN_",
  "appsecret_proof": "932e1d758c8379099e1b7f9e75e1abf41ab496760d64ddb05e3d21979d13c31f",
  "batch": "[{\\"method\\":\\"POST\\",\\"relative_url\\":\\"1234567890/messages?access_token=_ACCESS_TOKEN_\\",\\"body\\":\\"type=interactive&interactive=%7B%22type%22%3A%22buttons%22%2C%22body%22%3A%7B%22text%22%3A%22Hello%20*World!*%22%7D%2C%22header%22%3A%7B%22type%22%3A%22image%22%2C%22image%22%3A%7B%22link%22%3A%22http%3A%2F%2Fsociably.com%2Ffoo.jpg%22%7D%7D%2C%22actions%22%3A%7B%22buttons%22%3A%5B%7B%22type%22%3A%22reply%22%2C%22title%22%3A%22BAR%22%2C%22id%22%3A%22BAZ%22%7D%5D%7D%7D&to=9876543210&messaging_product=whatsapp\\",\\"name\\":\\"wa.1234567890.9876543210-1\\",\\"omit_response_on_success\\":false}]",
}
`;

exports[`#uploadMedia(message) post /media api 1`] = `
"----------------------------SEPERATOR
Content-Disposition: form-data; name=\\"file_0\\"
Content-Type: image/png

foo
----------------------------SEPERATOR
Content-Disposition: form-data; name=\\"access_token\\"

_ACCESS_TOKEN_
----------------------------SEPERATOR
Content-Disposition: form-data; name=\\"appsecret_proof\\"

932e1d758c8379099e1b7f9e75e1abf41ab496760d64ddb05e3d21979d13c31f
----------------------------SEPERATOR
Content-Disposition: form-data; name=\\"batch\\"

[{\\"method\\":\\"POST\\",\\"relative_url\\":\\"1234567890/media?access_token=_ACCESS_TOKEN_\\",\\"body\\":\\"type=image%2Fpng&messaging_product=whatsapp\\",\\"attached_files\\":\\"file_0\\",\\"omit_response_on_success\\":false}]
----------------------------SEPERATOR--
"
`;
